### ISSUE TITLE
Consider marking feature-dependent APIs with custom IDL Extended Attributes

### ISSUE THREAD

**Issue by toji** [gpuweb/gpuweb#1353]
TL;DR: We should consider explicitly annotating API surface that is feature-dependent, possibly with WebIDL's Extended Attribute syntax, like so:

```webidl
dictionary GPUDepthStencilState {
    required GPUTextureFormat format;
    // ...

    // Enable depth clamping (requires "depth-clamping" feature)
    [RequiresGPUFeature=depth-clamping] boolean clampDepth = false;
};
```

To aid developers that are referencing the spec to determine how to use a particular feature it would be very helpful if we had a consistent way of annotating that a given method, attribute, or dictionary value is only applicable if a specific feature has been enabled for the device. We should do this with comments at minimum, but it's also possible that we could use an existing feature of WebIDL for the annotations.

WebIDL [Extended Attributes](https://heycam.github.io/webidl/#idl-extended-attributes) are annotations that can be applied to types/members/etc of to control various aspects of the language bindings.  While the [most commonly used attributes](https://heycam.github.io/webidl/#es-extended-attributes) are part of the WebIDL core spec, [various](https://developer.mozilla.org/en-US/docs/Mozilla/WebIDL_bindings#Custom_extended_attributes) [browsers](https://chromium.googlesource.com/chromium/src/+/master/third_party/blink/renderer/bindings/IDLExtendedAttributes.md#Common-Blink_specific-IDL-Extended-Attributes) define their own extended attributes to control engine-specific behavior and specs have established a precedent of defining their extended attributes as well when needed. (See [[HTMLConstructor]](https://html.spec.whatwg.org/multipage/dom.html#htmlconstructor) or [[CEReactions]](https://html.spec.whatwg.org/multipage/custom-elements.html#cereactions)

Despite that, there may be some reasons why extended attributes would not be appropriate here. They are typically used to control how bindings are generated, and it's not clear that it would be desirable for the proposed attribute to actually change the bindings. (There is precedent for them also being used as simply hints for developers or implementors, however. For example: the [[SameObject]](https://chromium.googlesource.com/chromium/src/+/master/third_party/blink/renderer/bindings/IDLExtendedAttributes.md#SameObject_a) attribute has no effect on bindings generation in Blink.) Additionally, extended attributes cannot be applied to enums, so wouldn't be able to indicate enums that are only accepted when a given feature is enabled. Finally, if we anticipate having some interfaces that require multiple features to be enabled or any one of a number of features to be enabled we may find ourselves inventing a boolean syntax to try and express those requirements.

Still, I think it's worth evaluating as a way to make developers lives easier, and even if we determine that extended attributes aren't the right mechanism for it I'd like to see use come up with a consistent way of indicating feature requirements regardless.

**Comment by kvark** [comment]
The lack of enum variant support seems like a showstopper to me wrt this feature.
Note that for dictionaries we can also isolate the feature flags manually:
```
// Requires GPUFeature/"depth-clamping"
dictionary GPUDepthStencilStateForDepthClamping {
  boolean clampDepth = false;
};
dictionary GPUDepthStencilState: GPUDepthStencilStateForDepthClamping {
    required GPUTextureFormat format;
    // ...
};
```

**Comment by kainino0x** [comment]
We could even put the definition for that partial thing down in a separate section for the feature. However I wouldn't go that far as it makes it quite hard to discover what the actual list of members of a dictionary is.

I think I'd be okay with using only comments. But extended attributes would be pretty nice.

I find there's a fundamental tension with the way WebIDL+Bikeshed fit together, which is apparent here: WebIDL is not very central and is not treated like a "source of truth". For example if you add a definition for an idl member, then the idl itself becomes buried in the list of references, and it becomes hard to find the IDL from elsewhere. Meanwhile, we want our spec to be a useful reference to developers (like Vulkan, but unlike most Web specs), so the concise, discoverable nature of WebIDL makes it ideal as a primary reference point. But since that's not the case, maybe we need a style that replicates those properties of WebIDL, e.g. keeping our `<dl>`s as small as possible, maybe either putting algorithm steps outside of the `<dl>` or putting them in collapsible sections. Just some thoughts.

**Comment by toji** [comment]
> Note that for dictionaries we can also isolate the feature flags manually:

That doesn't feel like it scales past a single feature per dictionary, unless I'm missing something. Because we can't do multiple inheritance you'll be stuck in situations like this:

```webidl
// Requires feature 'foo'
dictionary GPUPipelineDescriptorWithFoo : GPUPipelineDescriptor {
  boolean foo = false;
}

// Requires feature 'bar'
dictionary GPUPipelineDescriptorWithBar : GPUPipelineDescriptor {
  unsigned long bar = 0;
}

partial interface GPUDevice {
   GPUPipeline createPipeline(/* Which goes here? */);
}
```

**Comment by kainino0x** [comment]
I think we would have to do it with partial dictionaries rather than dictionary inheritance.

**Comment by toji** [comment]
FYI: We got feedback in the WebXR group that partial dictionaries are a "thing that the WebIDL people regret" and were advised to avoid them. I don't recall that we were given a reason for that sentiment. 🤷

**Comment by kvark** [comment]
Editors discussion: this is interesting and useful, but we don't want to go this route yet.
